### PR TITLE
Fix change logging not being thread safe

### PR DESCRIPTION
**Pull request by gsnider2195**
# Closes: #3739
# What's Changed

Use [contextvars](https://docs.python.org/3/library/contextvars.html) to track the change logging state to make change logging thread safe.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/tests/test_context_managers.py [hunk 1]
```diff
@@ -1,8 +1,13 @@
+import threading
+import time
+
 from django.contrib.auth import get_user_model
 from django.contrib.contenttypes.models import ContentType
+from django.db import connection
 from django.test import TestCase
 
 from nautobot.core.celery import app
+from nautobot.utilities.testing import TransactionTestCase
```
Comment (gsnider2195) [inline nautobot/extras/tests/test_context_managers.py:10]: Need to move this down. Was initially trying to import from `nautobot.core`!

**Comment by gsnider2195** [comment]
Regarding the failing test, somehow the `nullify_connected_endpoints` signal is getting out of sync with the database in this branch but not in develop:

### develop:

preparing to delete cable
pre_delete signal for cable 0c6c10a3-ea6a-4e1c-a257-20ba5d4c0f53
Nullifying termination A for cable #0c6c10a3-ea6a-4e1c-a257-20ba5d4c0f53
circuittermination post_save signal received
deleting cable path 50370d50-27dc-42ae-9e0f-ad11c4c4c36b
save called for cable path **0ab7b030-2730-4969-8706-321e1ac6309a**
deleting cable path dce5ed6b-c772-43ae-a245-bf31b2eadc12
save called for cable path b4b5a91e-65c7-4a35-8b03-f91b91751709
deleting cable path e3e56eca-b320-418b-a0dd-4d084d11c78d
save called for cable path 6b12b488-9993-4d9d-bbba-4ecc427a236d
Nullifying termination B for cable #0c6c10a3-ea6a-4e1c-a257-20ba5d4c0f53
eth0 _path id is **0ab7b030-2730-4969-8706-321e1ac6309a**



### u/gas-3739-change-logging-thread-safe branch:

preparing to delete cable
pre_delete signal for cable cc731b78-198b-46fe-9db4-b60bdb3c64ad
Nullifying termination A for cable #cc731b78-198b-46fe-9db4-b60bdb3c64ad
circuittermination post_save signal received
deleting cable path **962d0614-7cec-474f-bc65-0ad7cc46efc8**
save called for cable path 62240c63-0f78-4d9d-85ae-b83636f406c6
deleting cable path 9b16293b-14a6-48b3-b988-d571c3c7e126
save called for cable path bd95b0b4-2430-4cfd-8377-00da8059064f
deleting cable path e0f2c695-15b3-42d5-9f4e-ccc5a9c59117
save called for cable path 4aa4cb36-740b-41b4-a88f-fd4c7e8d9e3d
Nullifying termination B for cable #cc731b78-198b-46fe-9db4-b60bdb3c64ad
eth0 _path id is **962d0614-7cec-474f-bc65-0ad7cc46efc8**


```py
@receiver(pre_delete, sender=Cable)
def nullify_connected_endpoints(instance, **kwargs):
    """
    When a Cable is deleted, check for and update its two connected endpoints
    """
    # Disassociate the Cable from its termination points
    print(f"pre_delete signal for cable {instance.pk}")
    if instance.termination_a is not None:
        print(f"Nullifying termination A for cable {instance}")
        instance.termination_a.cable = None
        instance.termination_a._cable_peer = None
        instance.termination_a.save()
    if instance.termination_b is not None:
        print(f"Nullifying termination B for cable {instance}")
        print(f"{instance.termination_b} _path id is {instance.termination_b._path_id}")
        instance.termination_b.cable = None
        instance.termination_b._cable_peer = None
        instance.termination_b.save()
```

The quick fix is to call `instance.termination_b.refresh_from_db()` before nulling out termination b but we should probably find the root cause before we merge.

**Review by timizuoebideri1 — APPROVED** [review]

File: nautobot/extras/signals.py [hunk 2]
```diff
@@ -70,29 +79,37 @@ def _handle_changed_object(change_context, sender, instance, raw=False, **kwargs
 
     # Record an ObjectChange if applicable
     if hasattr(instance, "to_objectchange"):
+        # save a copy of this instance's field cache so it can be restored after serialization
+        # to prevent unexpected behavior when chaining multiple signal handlers
+        original_cache = instance._state.fields_cache.copy()
```
Comment (glennmatthews) [inline nautobot/extras/signals.py:84]: Would it be preferable to fix the other handlers to not make assumptions about field caching? Not blocking.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/extras/signals.py (hunk 2 above)
Comment (gsnider2195) [reply nautobot/extras/signals.py:84]: Calling `instance.refresh_from_db()` does not refresh cached fields on cached related instances 😕

**Pull request merged**